### PR TITLE
fvm4 shim types

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:

- Implement shim types for `fvm@4`

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by LesnyRumcajs** [comment]
@hanabi1224 no green checkmark!

**Review by lemmih — APPROVED** [review]

**Comment by lemmih** [comment]
`ErrorNumber` can be radically simplified, but that's for another time.

**Comment by hanabi1224** [comment]
> @hanabi1224 no green checkmark!

@LesnyRumcajs Doctest fixed!

File: src/shim/clock.rs [hunk 1]
```diff
@@ -27,4 +27,12 @@ mod tests {
...
         )
     }
+
+    #[test]
+    fn fvm4_shim_of_const_epoch_duration_seconds() {
+        assert_eq!(
+            super::EPOCH_DURATION_SECONDS,
+            fvm_shared4::clock::EPOCH_DURATION_SECONDS
+        )
+    }
```
Comment (LesnyRumcajs) [inline src/shim/clock.rs:37]: This can be a static assertion instead of a test.

File: src/shim/crypto.rs [hunk 2]
```diff
@@ -88,7 +88,7 @@ impl Signature {
 
     /// Checks if a signature is valid given data and address.
     pub fn verify(&self, data: &[u8], addr: &crate::shim::address::Address) -> Result<(), String> {
-        use fvm_shared3::crypto::signature::ops::{verify_bls_sig, verify_secp256k1_sig};
+        use fvm_shared4::crypto::signature::ops::{verify_bls_sig, verify_secp256k1_sig};
```
Comment (LesnyRumcajs) [inline src/shim/crypto.rs:91]: should we rather use `fvm_shared_latest` here?

File: src/shim/gas.rs [hunk 3]
```diff
@@ -113,36 +128,62 @@ impl From<GasChargeV2> for GasCharge {
...
     fn from(value: GasCharge) -> Self {
-        GasChargeV2 {
+        Self {
             name: value.0.name,
-            compute_gas: GasV2::from_milligas(value.0.compute_gas.as_milligas() as i64),
-            storage_gas: GasV2::from_milligas(value.0.other_gas.as_milligas() as i64),
+            compute_gas: GasV3::from_milligas(value.0.compute_gas.as_milligas() as _),
+            other_gas: GasV3::from_milligas(value.0.other_gas.as_milligas() as _),
+            // FIXME: https://github.com/ChainSafe/forest/issues/3524
+            elapsed: Default::default(),
```
Comment (LesnyRumcajs) [inline src/shim/gas.rs:164]: can we fix it now?

File: src/tool/subcommands/snapshot_cmd.rs [hunk 4]
```diff
@@ -24,7 +24,7 @@ use clap::Subcommand;
 use dialoguer::{theme::ColorfulTheme, Confirm};
 use futures::TryStreamExt;
 use fvm_ipld_blockstore::Blockstore;
-use fvm_shared3::address::Network;
+use fvm_shared4::address::Network;
```
Comment (LesnyRumcajs) [inline src/tool/subcommands/snapshot_cmd.rs:27]: Perhaps we could export `latest` from this shim to have this versioning mostly contained there.

File: src/shim/gas.rs (hunk 3 above)
Comment (hanabi1224) [reply src/shim/gas.rs:164]: Unfortunately `GasDuration` is not exported from fvm@3, I'm thinking about creating a macro that implements `From` with unsafe `std::mem` API to simplify type conversions between fvm@3 and fvm@4 given most of the type pairs are identical between them

File: src/shim/clock.rs (hunk 1 above)
Comment (hanabi1224) [reply src/shim/clock.rs:37]: Do you mean using `const_assert_eq!`? Where should I put the assertion code?

File: src/shim/crypto.rs (hunk 2 above)
Comment (hanabi1224) [reply src/shim/crypto.rs:91]: Fixed.

File: src/tool/subcommands/snapshot_cmd.rs (hunk 4 above)
Comment (hanabi1224) [reply src/tool/subcommands/snapshot_cmd.rs:27]: Fixed.

File: src/shim/clock.rs (hunk 1 above)
Comment (LesnyRumcajs) [reply src/shim/clock.rs]: I removed our own re-definition of `EPOCH_DURATION_SECONDS` and just re-exported it from the FVM4 + removed the tests.

**Pull request merged**